widget/calendar.go

```diff
@@ -0,0 +1,213 @@
+
+	dateContainer := &fyne.Container{Layout: layout.NewBorderLayout(nav, nil, nil, nil),
+		Objects: []fyne.CanvasObject{nav, c.dates}}
+
+	return NewSimpleRenderer(dateContainer)
+}
+
+func (c *Calendar) calendarObjects() []fyne.CanvasObject {
+	var columnHeadings []fyne.CanvasObject
+	for i := 0; i < daysPerWeek; i++ {
+		j := i + 1
+		if j == daysPerWeek {
+			j = 0
+		}
+
+		t := NewLabel(strings.ToUpper(time.Weekday(j).String()[:3]))
+		t.Alignment = fyne.TextAlignCenter
+		columnHeadings = append(columnHeadings, t)
+	}
+	columnHeadings = append(columnHeadings, c.daysOfMonth()...)
+
+	return columnHeadings
+}
+
+func (c *Calendar) dateForButton(dayNum int) time.Time {
+	oldName, off := c.currentTime.Zone()
+	return time.Date(c.currentTime.Year(), c.currentTime.Month(), dayNum, c.currentTime.Hour(), c.currentTime.Minute(), 0, 0, time.FixedZone(oldName, off)).In(c.currentTime.Location())
+}
+
+func (c *Calendar) daysOfMonth() []fyne.CanvasObject {
+	start := time.Date(c.currentTime.Year(), c.currentTime.Month(), 1, 0, 0, 0, 0, c.currentTime.Location())
+	var buttons []fyne.CanvasObject
+
+	//account for Go time pkg starting on sunday at index 0
+	dayIndex := int(start.Weekday())
+	if dayIndex == 0 {
+		dayIndex += daysPerWeek
+	}
+
+	//add spacers if week doesn't start on Monday
+	for i := 0; i < dayIndex-1; i++ {
+		buttons = append(buttons, layout.NewSpacer())
+	}
+
+	for d := start; d.Month() == start.Month(); d = d.AddDate(0, 0, 1) {
+
+		dayNum := d.Day()
+		s := strconv.Itoa(dayNum)
+		b := NewButton(s, func() {
+
+			selectedDate := c.dateForButton(dayNum)
+
+			c.OnChanged(selectedDate)
+		})
+		b.Importance = LowImportance
+
+		buttons = append(buttons, b)
+	}
+
+	return buttons
+}
+
+func (c *Calendar) monthYear() string {
+	return c.currentTime.Format("January 2006")
+}
+
+type calendarLayout struct {
+	cellSize fyne.Size
+}
+
+func newCalendarLayout() fyne.Layout {
+	return &calendarLayout{}
+}
+
+// Layout is called to pack all child objects into a specified size.
+// For a calendar grid this will pack objects into a table format.
+func (g *calendarLayout) Layout(objects []fyne.CanvasObject, size fyne.Size) {
+	weeks := 1
+	day := 0
+	for i, child := range objects {
+		if !child.Visible() {
+			continue
+		}
+
+		if day%daysPerWeek == 0 && i >= daysPerWeek {
+			weeks++
+		}
+		day++
+	}
+
+	g.cellSize = fyne.NewSize(size.Width/float32(daysPerWeek),
+		size.Height/float32(weeks))
+	row, col := 0, 0
+	i := 0
+	for _, child := range objects {
+		if !child.Visible() {
+			continue
+		}
+
+		lead := g.getLeading(row, col)
+		trail := g.getTrailing(row, col)
+		child.Move(lead)
+		child.Resize(fyne.NewSize(trail.X, trail.Y).Subtract(lead))
+
+		if (i+1)%daysPerWeek == 0 {
+			row++
+			col = 0
+		} else {
+			col++
+		}
+		i++
+	}
+}
+
+// MinSize sets the minimum size for the calendar
+func (g *calendarLayout) MinSize(_ []fyne.CanvasObject) fyne.Size {
+	pad := theme.Padding()
+	largestMin := NewLabel("22").MinSize()
```

Comment:
I think this template label can be declared once as a private global, to avoid allocating new widgets every MinSize?